web/src/main/java/org/springframework/security/web/authentication/rememberme/TokenBasedRememberMeServices.java

```diff
@@ -51,13 +51,20 @@
  * This is also necessary so that the user's password is available and can be checked as
  * part of the encoded cookie.
  * <p>
- * The cookie encoded by this implementation adopts the following form:
+ * The cookie encoded by this implementation adopts either the following form:
  *
  * <pre>
  * username + &quot;:&quot; + expiryTime + &quot;:&quot;
  * 		+ Md5Hex(username + &quot;:&quot; + expiryTime + &quot;:&quot; + password + &quot;:&quot; + key)
  * </pre>
  *
+ * or, if a hashing algorithm is configured, the form:
+ *
+ * <pre>
+ * username + &quot;:&quot; + expiryTime + &quot;:&quot; + algorithmName + &quot;:&quot;
```

Comment:
I wonder if this should be thought of in the same way as how Spring Security deals with passwords.

For example, you might consider having an algorithm-for-matches and an algorithm-for-encoding, both defaulting to MD5.

* When encoding, encode with the algorithm-for-encoding, adding the prefix either way (even if the app hasn't configured it directly)
* When decoding
** if there is no algorithm in the cookie, use the algorithm-for-matches
** if there is an algorithm in the cookie, use that algorithm

This will allow folks to safely upgrade to SHA-256 without losing their old MD5 hash cookies.